boot/assets.go

```diff
@@ -790,16 +794,40 @@ func observeSuccessfulBootAssetsForBootloader(m *Modeenv, root string, opts *boo
 		return nil, nil
 	}
 
-	cache := newTrustedAssetsCache(dirs.SnapBootAssetsDir)
-	for _, trustedAsset := range trustedAssets {
+	byAssetName := make(map[string][]string)
+	for trustedAsset := range trustedAssets {
 		assetName := filepath.Base(trustedAsset)
+		l, ok := byAssetName[assetName]
+		if !ok {
+			byAssetName[assetName] = []string{trustedAsset}
+		} else {
+			byAssetName[assetName] = append(l, trustedAsset)
+		}
+	}
 
-		// find the hash of the file on disk
-		assetHash, err := cache.fileHash(filepath.Join(root, trustedAsset))
-		if err != nil && !os.IsNotExist(err) {
-			return nil, fmt.Errorf("cannot calculate the digest of existing trusted asset: %v", err)
+	cache := newTrustedAssetsCache(dirs.SnapBootAssetsDir)
+
+	for assetName, trustedAssetsForName := range byAssetName {
```

Comment:
trustedAssets is not sorted and this code remembers only one asset, isn't there a chance that we will find exactly the one that we didn't remember?